### PR TITLE
mgr/volumes: Show clone progress

### PR DESCRIPTION
**Pull request by kotreshhr**
Add capability to show clone progress and failure reason in clone status command.

The clone progress is shown with following.
    
     1. size                : Size of the clone
     2. percentage          : Progress of the clone in percentage

    
Note: The progress shown in not accurate. The 'size' depends on recursive
stat which gets updated at intervals. The same applies for 'percentage'.
    
    The sample output of clone status in 'in-progress' state:
      $ ceph fs clone status cephfs clone1
      {
        "status": {
          "state": "in-progress",
          "source": {
            "volume": "cephfs",
            "subvolume": "subvol1",
            "snapshot": "snap1"
            "size": "104857600"
          },
          "progress": {
            "size": "8388608",
            "percentage": "8.00",
          },
          "failure": {
            "errno": "122",
            "errstr": "Disk quota exceeded"
          }
    
    The sample output from an `pending` clone operation::
      $ ceph fs clone status cephfs clone1
      {
        "status": {
          "state": "pending",
          "source": {
            "volume": "cephfs",
            "subvolume": "subvol1",
            "snapshot": "snap1"
            "size": "104857600"
          },
          "progress": {
            "size": "0",
            "percentage": "0.00",
          },
           "failure": {
            "errno": "122",
            "errstr": "Disk quota exceeded"
          }
        }
      }
    
Fixes: https://tracker.ceph.com/issues/53949
Signed-off-by: Kotresh HR <khiremat@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by kotreshhr** [comment]
jenkins test make check

**Comment by kotreshhr** [comment]
jenkins render docs

**Comment by kotreshhr** [comment]
jenkins test make check

**Comment by kotreshhr** [comment]
jenkins test make check

**Comment by kotreshhr** [comment]
jenkins render docs

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 1]
```diff
@@ -660,6 +690,44 @@ def _get_clone_source(self):
...
+            except MetadataMgrException as me:
+                if me.errno == -errno.ENOENT:
+                    clone_started = False
+                else:
+                    raise
+
+            clone_usedbytes = int(self._recursive_size(self.path))
+            progress = "0.00" if src_usedbytes == 0 else '{0:.2f}'.format((clone_usedbytes / src_usedbytes) * 100.0)
+
+            # Calculate approximate pending time for clone. The time of beginning of bulk copy is considered
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:716]: Agree, I will take out the pending time.

File: src/pybind/mgr/volumes/fs/async_cloner.py [hunk 2]
```diff
@@ -200,13 +201,31 @@ def set_quota_on_clone(fs_handle, clone_volumes_pair):
...
     with open_volume_lockless(fs_client, volname) as fs_handle:
         with open_clone_subvolume_pair(fs_client, fs_handle, volspec, volname, groupname, subvolname) as clone_volumes:
             src_path = clone_volumes[1].snapshot_data_path(clone_volumes[2])
             dst_path = clone_volumes[0].path
+            start_time = datetime.now()
+            update_clone_progress(fs_handle, clone_volumes, start_time)
             bulk_copy(fs_handle, src_path, dst_path, should_cancel)
             set_quota_on_clone(fs_handle, clone_volumes)
+            end_time = datetime.now()
+            # start_time could have changed if crashed and recovered
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/async_cloner.py:224]: No, it doesn't get removed before restarting the clone. I meant `start_time` at line 219 is re-initialized which is in-correct. So fetching it from config file.

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 3]
```diff
@@ -660,6 +690,44 @@ def _get_clone_source(self):
...
+                approx_pending_time = str(timedelta(seconds=int(pending_time)))
+            else:
+                approx_pending_time = "indefinite"
+
+            clone_progress = {
+                    'size'               : str(clone_usedbytes),
+                    'percentage'         : progress,
+                    'approx_pending_time': approx_pending_time,
+            }
+        except cephfs.Error as e:
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:732]: Ack, will do it.

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 4]
```diff
@@ -139,6 +143,38 @@ def remove_clone_source(self, flush=False):
...
+                self.metadata_mgr.update_section("progress", "start_time", start_time)
+            if flush:
+                self.metadata_mgr.flush()
+
+    def remove_clone_progress(self, flush=False):
+        self.metadata_mgr.remove_section("progress")
+        if flush:
+            self.metadata_mgr.flush()
+
+    def add_clone_failure(self, errno, errstr, flush=False):
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:171]: Sorry, forgot to add the file while committing! I will add the changes.

[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:171]: Done

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py]: Done

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py]: Done

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 5]
```diff
@@ -716,7 +730,15 @@ def status(self):
         }
         if not SubvolumeOpSm.is_complete_state(state) and subvolume_type == SubvolumeTypes.TYPE_CLONE:
             subvolume_status["source"] = self._get_clone_source()
-            subvolume_status["progress"] = self._get_clone_progress()
+            try:
+                subvolume_status["progress"] = self._get_clone_progress()
+            except VolumeException:
+                pass
+        if SubvolumeOpSm.is_failed_state(state) and subvolume_type == SubvolumeTypes.TYPE_CLONE:
```
Comment (kotreshhr) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:737]: @ajarr @vshankar Should we treat `canceled` state as failure? If so it  might need a failure status update while transitioning from `pending` to `canceled`.

File: src/pybind/mgr/volumes/fs/async_cloner.py [hunk 6]
```diff
@@ -217,13 +218,31 @@ def set_quota_on_clone(fs_handle, clone_volumes_pair):
...
             dst_path = clone_volumes[0].path
+            start_time = datetime.now()
+            update_clone_progress(fs_handle, clone_volumes, start_time)
             bulk_copy(fs_handle, src_path, dst_path, should_cancel)
             set_quota_on_clone(fs_handle, clone_volumes)
+            end_time = datetime.now()
+            # start_time could have changed if crashed and recovered. Hence fetch from config file
+            actual_start_time = datetime.fromisoformat(clone_volumes[0].metadata_mgr.get_option("progress", "start_time"))
+            idle_time = clone_volumes[0].metadata_mgr.get_option("progress", "idle_time")
+            log.info("Clone copy time: ({0}, {1}, {2}, idle_time = {3}, completed_time = {4})".format(volname,
```
Comment (mchangir) [inline src/pybind/mgr/volumes/fs/async_cloner.py:244]: if possible, please use `f""` format strings ... they are more intuitive to understand

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 7]
```diff
@@ -660,6 +690,23 @@ def _get_clone_source(self):
...
+    def _get_clone_progress(self):
+        try:
+            src_snappath = self.metadata_mgr.get_option("source", "snappath")
+            src_usedbytes = int(self.fs.getxattr(src_snappath, 'ceph.dir.rbytes'))
+            clone_usedbytes = int(self._recursive_size(self.path))
+        except MetadataMgrException as me:
+            raise VolumeException(-errno.EINVAL, "error fetching clone progress metadata - {0}".format(me))
+        except cephfs.Error as e:
+            raise VolumeException(-e.args[0], "error fetching clone progress metadata - {0}".format(e.args[1]))
+
```
Comment (mchangir) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:709]: if possible, please use `f""` format strings ... they are more intuitive to understand

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 8]
```diff
@@ -660,6 +690,23 @@ def _get_clone_source(self):
...
+        progress = "0.00" if src_usedbytes == 0 else '{0:.2f}'.format((clone_usedbytes / src_usedbytes) * 100.0)
```
Comment (mchangir) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:710]: if possible, please use `f""` format strings ... they are more intuitive to understand

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py (hunk 5 above)
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:740]: The code is updated to display the failure status when the clone is in `failed` and `canceled` state. When in clone is cancelled, failure error `EINTR` is shown. I hope that is fine.

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:740]: For the user canceled case, lets show "User Interrupted" in the error string (relying on `strerror(EINTR)` could be misleading)

Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:740]: Done. I have chosen "user interrupted clone operation"

**Comment by gregsfortytwo** [comment]
Jenkins test readthedocs

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 9]
```diff
@@ -139,6 +143,31 @@ def remove_clone_source(self, flush=False):
         if flush:
             self.metadata_mgr.flush()
 
+    def add_clone_progress(self, idle_time, start_time, flush=False):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:146]: don't you always want to flush?

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 10]
```diff
@@ -139,6 +143,31 @@ def remove_clone_source(self, flush=False):
...
+            self.metadata_mgr.get_option("progress", "idle_time")
+            self.metadata_mgr.get_option("progress", "start_time")
+        except MetadataMgrException as me:
+            if me.errno == -errno.ENOENT:
+                self.metadata_mgr.update_section("progress", "idle_time", idle_time)
+                self.metadata_mgr.update_section("progress", "start_time", start_time)
+            if flush:
+                self.metadata_mgr.flush()
+
+    def remove_clone_progress(self, flush=False):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:166]: don't you always want to flush?

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 11]
```diff
@@ -660,6 +690,23 @@ def _get_clone_source(self):
             raise VolumeException(-errno.EINVAL, "error fetching subvolume metadata")
         return clone_source
 
+    def _get_clone_progress(self):
+        try:
+            src_snappath = self.metadata_mgr.get_option("source", "snappath")
+            src_usedbytes = int(self.fs.getxattr(src_snappath, 'ceph.dir.rbytes'))
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:703]: I guess you could also use, self.metadata_mgr.get_option("source", "size")?

**Comment by ajarr** [comment]
Tests look comprehensive. Nice!

File: qa/tasks/cephfs/test_volumes.py [hunk 12]
```diff
@@ -3903,6 +3914,97 @@ def test_subvolume_snapshot_attr_clone(self):
...
+
+        # schedule a clone1
+        self._fs_cmd("subvolume", "snapshot", "clone", self.volname, subvolume, snapshot, clone1)
+
+        # check clone1 to be in-progress
+        self._wait_for_clone_to_be_in_progress(clone1)
+
+        # check clone1 progress status
+        clone1_result = self._get_clone_status(clone1)
+        self.assertLessEqual(int(clone1_result["status"]["progress"]["size"]), int(clone1_result["status"]["source"]["size"]))
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply qa/tasks/cephfs/test_volumes.py:3945]: I have used assertLessEqual, so should be fine but makes sense to add one assert after complete as well. I will add that.

[the remark this answers is not in the document]
Comment (kotreshhr) [reply qa/tasks/cephfs/test_volumes.py:3945]: Nope, the clone progress is not shown on a `completed` clone.

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 13]
```diff
@@ -168,6 +168,13 @@ def remove_clone_progress(self, flush=False):
         if flush:
             self.metadata_mgr.flush()
 
+    def add_clone_failure(self, errno, errstr, flush=False):
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py]: Removed optional flush

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py (hunk 10 above)
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py]: Removed optional flush

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py (hunk 11 above)
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py]: Done

**Comment by kotreshhr** [comment]
@vshankar, @ceph/cephfs 

Because of the recursive rstats issue on snapshot path [1], the clone-progress is incorrect and shows 0 for clones from 'retained-snapshot' subvolume snapshots. We can workaround this in mgr/volumes by storing the snapshot size in the metadata file. But doing so would still be buggy, if the recursive rstats, which is lazy and not updated for subvolumes at the time of taking the snapshot. Is this acceptable ?


[1] https://tracker.ceph.com/issues/55108

**Comment by vshankar** [comment]
> @vshankar, @ceph/cephfs
> 
> Because of the recursive rstats issue on snapshot path [1], the clone-progress is incorrect and shows 0 for clones from 'retained-snapshot' subvolume snapshots. We can workaround this in mgr/volumes by storing the snapshot size in the metadata file. But doing so would still be buggy, if the recursive rstats, which is lazy and not updated for subvolumes at the time of taking the snapshot. Is this acceptable ?
> 
> [1] https://tracker.ceph.com/issues/55108

Yes. Let's revisit this when we have layering support for CephFS snapshots.

**Comment by kotreshhr** [comment]
> > @vshankar, @ceph/cephfs
> > Because of the recursive rstats issue on snapshot path [1], the clone-progress is incorrect and shows 0 for clones from 'retained-snapshot' subvolume snapshots. We can workaround this in mgr/volumes by storing the snapshot size in the metadata file. But doing so would still be buggy, if the recursive rstats, which is lazy and not updated for subvolumes at the time of taking the snapshot. Is this acceptable ?
> > [1] https://tracker.ceph.com/issues/55108
> 
> Yes. Let's revisit this when we have layering support for CephFS snapshots.

Makes sense, closing this for now.

**Pull request closed without merging**